### PR TITLE
[SPARK-19851] Add support for EVERY and ANY (SOME) aggregates

### PR DESCRIPTION
**Pull request by dilipbiswal**
## What changes were proposed in this pull request?
This PR is a rebased version of original work [link](https://github.com/apache/spark/pull/17648) by 
@ptkool. 

Please give credit to @ptkool for this work.

Description from original PR:
This pull request implements the EVERY and ANY aggregates.

## How was this patch tested?
Testing was performed using unit tests, integration tests, and manual tests.

### CODE REVIEW

**Comment by dilipbiswal** [comment]
@gatorsmile I tried to implement the rewrites suggested in the original PR. It does not seem very straightforward to me. The basic issue is, we are unable to replace the aggregate expression  to a scalar expression over aggregates. We only support limited number of true aggregate expressions under window.

For example -we are unable to rewrite .
```
select key, value, some(value) over(partition by key order by value) from src group by key, value
```
to
```
select key, value, coalesce(max(c1) == true, false) over(partition by key order by value) from src group by key, value
```
I tried a similar frame work to replace aggregate expressions like `ReplaceExpressions`. Please let me know what you think.

**Comment by HyukjinKwon** [comment]
> Please give credit to @ptkool for this work.

FWIW, we can now credit to multiple people per https://github.com/apache/spark/commit/51bee7aca13451167fa3e701fcd60f023eae5e61 :-)

File: python/pyspark/sql/functions.py [hunk 1]
```diff
@@ -202,6 +202,12 @@ def _():
                """,
 }
 
+_functions_2_2 = {
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: hm, looks unrelated.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 2]
```diff
@@ -1555,9 +1555,11 @@ case class Left(str: Expression, len: Expression, child: Expression) extends Run
  * number of bytes of the given binary expression.
  */
 @ExpressionDescription(
-  usage = "_FUNC_(expr) - Returns the character length of string data or number of bytes of " +
-    "binary data. The length of string data includes the trailing spaces. The length of binary " +
-    "data includes binary zeros.",
+  usage = """
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: Looks unrelated

File: sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala [hunk 3]
```diff
@@ -88,7 +88,7 @@ class RelationalGroupedDataset protected[sql](
   }
 
   private[this] def aggregateNumericColumns(colNames: String*)(f: Expression => AggregateFunction)
-    : DataFrame = {
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala]: nit: previous indentation was correct.

File: sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala [hunk 4]
```diff
@@ -297,9 +318,44 @@ class RelationalGroupedDataset protected[sql](
   }
 
   /**
-   * Pivots a column of the current `DataFrame` and performs the specified aggregation.
+   * Compute the logical and of all boolean columns for each group.
+   * The resulting `DataFrame` will also contain the grouping columns.
+   * When specified columns are given, only compute the sum for them.
+   *
+   * @since 2.2.0
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala]: nit: since version

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala (hunk 2 above)
Comment (dilipbiswal) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: @HyukjinKwon Not sure why.. when i did a build/sbt doc , i got an error here. Thats the reason i had to fix.

File: sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala (hunk 4 above)
Comment (dilipbiswal) [reply sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala]: Thanks. will fix

File: sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala (hunk 3 above)
Comment (dilipbiswal) [reply sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala]: @HyukjinKwon Thanks ... will fix.

File: python/pyspark/sql/functions.py (hunk 1 above)
Comment (dilipbiswal) [reply python/pyspark/sql/functions.py]: @HyukjinKwon Will look into this.

**Comment by dilipbiswal** [comment]
retest this please

**Comment by gatorsmile** [comment]
Let me post something I wrote recently. Could you add test cases to ensure that we do not break the  "Ignore NULLs" policy

> All the set/aggregate functions ignore NULLs. The typical built-in Set/Aggregate functions are AVG, COUNT, MAX, MIN, SUM, GROUPING. 

> Note, COUNT(*) is actually equivalent to COUNT(1). Thus, it still includes rows containing null.

> Tip, because of the "Ignore NULLs" policy, Sum(a) + Sum(b) is not the same as Sum(a+b). 

> Note, although the set functions follow the "Ignore NULLs" policy, MIN, MAX, SUM AVG, EVERY, ANY and SOME returns NULL if 1) every value is NULL or 2) SELECT returns no row at all. COUNT never returns NULL. 

> TODO: When a set function eliminates NULLs, Spark SQL does not follow others to issue a warning message SQLSTATE 01003 "null value eliminated in set function".

> TODO: Check whether all the expressions that extend AggregateFunction follow the "Ignore NULLs" policy. If not, we need more investigation to see whether we should correct them.

> TODO: When Spark SQL supports EVERY, ANY, and SOME, they follow the same "Ignore NULLs" policy.

**Comment by dilipbiswal** [comment]
@gatorsmile Thanks.. I will check.

**Comment by dilipbiswal** [comment]
@gatorsmile First of all, thank you very much . Actually the added aggregates weren't null filtering. I have fixed the issue and have added additional test cases. Thank you.

**Comment by cloud-fan** [comment]
is it possible to rewrite these 3 new functions with existing expression? e.g.
`every(col)` -> `count(if (col) null else 1) == 0`
`any(col)` -> `count(if (col) 1 else null) > 0`

**Comment by dilipbiswal** [comment]
@cloud-fan please see my comment [link](https://github.com/apache/spark/pull/22047#issuecomment-411536039). I had tried to rewrite using max and min as suggested by Herman and Reynold in the original pr. I was unable to do it when the aggregate is part of the window.

File: python/pyspark/sql/functions.py [hunk 5]
```diff
@@ -403,6 +403,28 @@ def countDistinct(col, *cols):
     return Column(jc)
 
 
+def every(col):
```
Comment (gatorsmile) [inline python/pyspark/sql/functions.py]: Please keep the SQL functions and remove the function APIs. Thanks!

Comment (dilipbiswal) [reply python/pyspark/sql/functions.py]: @gatorsmile OK.

Comment (dilipbiswal) [reply python/pyspark/sql/functions.py]: @gatorsmile Hi Sean, I have prepared two branches. One in which these new aggregate functions are extending from the base Max and Min class basically reusing code. The other in which we replace these aggregate expressions in the optimizer. Below are the links.

1. [branch-extend](https://github.com/dilipbiswal/spark/tree/SPARK-19851-extend)

2. [branch-rewrite](https://github.com/dilipbiswal/spark/tree/SPARK-19851-rewrite)

I would prefer option 1 because of the following reasons.
1. Code changes are simpler
2. Supports these aggregates as window expressions naturally. In the other option i have
    to block it. 
3. It seems to me for these simple mapping, we probably don't need a rewrite frame work. We could add it in the future if we need a little complex transformation.

Please let me know how we want to move forward with this. Thanks !!

Comment (cloud-fan) [reply python/pyspark/sql/functions.py]: +1 for option 1

Comment (dilipbiswal) [reply python/pyspark/sql/functions.py]: @cloud-fan Thank you very much for your response. I will create a new PR based on option-1 today and close this one.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/AnyAgg.scala [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+package org.apache.spark.sql.catalyst.expressions.aggregate
+
+import org.apache.spark.sql.catalyst.analysis.TypeCheckResult
+import org.apache.spark.sql.catalyst.dsl.expressions._
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.util.TypeUtils
+import org.apache.spark.sql.types._
+
+@ExpressionDescription(
+  usage = "_FUNC_(expr) - Returns true if at least one value of `expr` is true.")
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/AnyAgg.scala]: BTW, don't forget to add `since`.
